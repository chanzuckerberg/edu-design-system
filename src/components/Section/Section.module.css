@import '../../design-tokens/mixins.css';

/*------------------------------------*\
    #SECTION
\*------------------------------------*/

/**
 * Section
 * 1) A section is a discrete section of a web page
 */
.section {
  padding-top: var(--eds-size-4);
  padding-bottom: var(--eds-size-4);
}

/**
 * Section header
 * 1) contains the section title, description and other header content
 */
.section__header {
  display: flex;
  flex-direction: column;
  margin-bottom: var(--eds-size-3);

  /**
     * Section header within centered section
     */
  .section--center & {
    align-items: center;
  }
}

/**
 * Section header
 * 1) contains the section title, description and other header content
 */
.section__header-inner {
  display: flex;
  align-items: center;
  gap: var(--eds-size-2);
}

/**
 * Section title
 */
.section__title {
  margin: 0;
  @mixin eds-theme-typography-headline-md;
}

/**
 * Section description
 */
.section__description {
  @mixin eds-theme-typography-body-text-md;
  @mixin reset;
  margin-top: var(--eds-size-1);
}
