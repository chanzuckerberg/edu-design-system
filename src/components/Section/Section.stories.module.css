@import '../../design-tokens/mixins.css';

/**
 *  Body text xs bold
 */
.section__text {
  @mixin eds-theme-typography-body-xs-bold;
}
