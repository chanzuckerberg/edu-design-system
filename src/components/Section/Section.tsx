import clsx from 'clsx';
import type { ReactNode } from 'react';
import React from 'react';
import styles from './Section.module.css';
import type { HeadingElement, HeadingSize } from '../Heading';
import Heading from '../Heading';
export interface Props {
  /**
   * Align variations:
   * - **center** yields a center-aligned section header
   */
  align?: 'center';
  /**
   * Child node(s) that can be nested inside component
   */
  children?: ReactNode;
  /**
   * CSS class names that can be appended to the component.
   */
  className?: string;
  /**
   * Description that appears below the section title
   */
  description?: ReactNode;
  /**
   * overline appears above the section title
   */
  overline?: ReactNode;
  /**
   * "as" prop, passed to Heading Component
   */
  headingAs?: HeadingElement;
  /**
   * "size" prop, passed to Heading Component
   */
  headingSize?: HeadingSize;
  /**
   * Right slot - an area to put right-aligned content after section title
   */
  right?: ReactNode;
  /**
   * Section heading text string
   */
  title?: string;
  /**
   * Slot for node to appear to the right of the section title. Typically used to include a Badge, Button, Tooltip, or other component
   */
  titleAfter?: ReactNode;
  /**
   * Slot for node to appear to the left of the section title. Typically used for images or avatars
   */
  titleBefore?: ReactNode;
}

/**
 * `import {Section} from "@chanzuckerberg/eds";`
 *
 * Section component contains a section header and body.
 *
 * The Heading component requires a value for "size", so this headingAs prop is provided
 * a default value of "h2" to allow it to remain optional on Section component.
 */
export const Section = ({
  align,
  children,
  className,
  description,
  headingAs,
  headingSize = 'h2',
  overline,
  right,
  title,
  titleAfter,
  titleBefore,
  ...other
}: Props) => {
  const componentClassName = clsx(
    styles['section'],
    align === 'center' && styles['section--center'],
    className,
  );

  return (
    <section className={componentClassName} {...other}>
      <div className={styles['section__inner']}>
        <header className={styles['section__header']}>
          <div className={styles['section__header-inner']}>
            {titleBefore && (
              <div className={styles['section__title-before']}>
                {titleBefore}
              </div>
            )}
            <div className={styles['section__title-inner']}>
              {overline && (
                <div className={styles['section__overline']}>{overline}</div>
              )}
              <Heading
                as={headingAs}
                className={styles['section__title']}
                size={headingSize}
              >
                {title}
                {titleAfter && (
                  <span className={styles['section__title-after']}>
                    {titleAfter}
                  </span>
                )}
              </Heading>
              {description && (
                <div className={styles['section__description']}>
                  {description}
                </div>
              )}
            </div>
          </div>
          {right && <div className={styles['section__right']}>{right}</div>}
        </header>
        <div className={styles['section__body']}>{children}</div>
      </div>
    </section>
  );
};
