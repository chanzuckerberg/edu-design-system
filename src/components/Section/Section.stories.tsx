import { BADGE } from '@geometricpanda/storybook-addon-badges';
import type { StoryObj, Meta } from '@storybook/react';
import React from 'react';
import { Section } from './Section';
import Button from '../Button';
import Icon from '../Icon';
import Text from '../Text';

export default {
  title: 'Components/Section',
  component: Section,
  parameters: {
    badges: ['1.0', BADGE.DEPRECATED],
  },
  args: {
    children:
      'This is the section body, where you can put any content or include other components.',
  },
} as Meta<Args>;

type Args = React.ComponentProps<typeof Section>;

const HelpIcon = (
  <Icon name="help" purpose="informative" size="1.375rem" title="help" />
);

export const Default: StoryObj<Args> = {
  args: {
    title: 'Section Title',
  },
};

export const Center: StoryObj<Args> = {
  args: {
    align: 'center',
    overline: <Text preset="overline">Overline above title</Text>,
    title: 'Section Title',
    description: (
      <Text as="p" preset="body-xs">
        This is a description of what the section is
      </Text>
    ),
  },
};

export const WithDescription: StoryObj<Args> = {
  args: {
    title: 'Section Title',
    description: (
      <Text as="p" preset="body-xs">
        This is a description of what the section is
      </Text>
    ),
  },
};

export const WithOverline: StoryObj<Args> = {
  args: {
    overline: <Text preset="overline">Overline above title</Text>,
    title: 'Section Title',
  },
};

export const WithRight: StoryObj<Args> = {
  args: {
    right: (
      <Button size="md" variant="primary">
        Button
      </Button>
    ),
    title: 'Section Title',
  },
};

export const WithTitleAfter: StoryObj<Args> = {
  args: {
    titleAfter: HelpIcon,
    title: 'Section Title',
  },
};

export const WithTitleBefore: StoryObj<Args> = {
  args: {
    titleBefore: HelpIcon,
    title: 'Section Title',
    headingAs: 'h4',
  },
};
