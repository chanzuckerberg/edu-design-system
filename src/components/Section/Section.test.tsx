import { generateSnapshots } from '@chanzuckerberg/story-utils';
import * as stories from './Section.stories';

describe('<Section />', () => {
  generateSnapshots(stories);
});
