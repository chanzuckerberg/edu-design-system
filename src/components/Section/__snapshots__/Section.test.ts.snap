// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Section /> Center story renders snapshot 1`] = `
<section
  class="section section--center"
>
  <div
    class="section__inner"
  >
    <header
      class="section__header"
    >
      <div
        class="section__header-inner"
      >
        <div
          class="section__title-inner"
        >
          <div
            class="section__overline"
          >
            <p
              class="text text--overline"
            >
              Overline above title
            </p>
          </div>
          <h2
            class="heading heading--size-h2 section__title"
          >
            Section Title
          </h2>
          <div
            class="section__description"
          >
            <p
              class="text text--sm section__text"
            >
              This is a description of what the section is
            </p>
          </div>
        </div>
      </div>
    </header>
    <div
      class="section__body"
    >
      This is the section body, where you can put any content or include other components.
    </div>
  </div>
</section>
`;

exports[`<Section /> Default story renders snapshot 1`] = `
<section
  class="section"
>
  <div
    class="section__inner"
  >
    <header
      class="section__header"
    >
      <div
        class="section__header-inner"
      >
        <div
          class="section__title-inner"
        >
          <h2
            class="heading heading--size-h2 section__title"
          >
            Section Title
          </h2>
        </div>
      </div>
    </header>
    <div
      class="section__body"
    >
      This is the section body, where you can put any content or include other components.
    </div>
  </div>
</section>
`;

exports[`<Section /> WithDescription story renders snapshot 1`] = `
<section
  class="section"
>
  <div
    class="section__inner"
  >
    <header
      class="section__header"
    >
      <div
        class="section__header-inner"
      >
        <div
          class="section__title-inner"
        >
          <h2
            class="heading heading--size-h2 section__title"
          >
            Section Title
          </h2>
          <div
            class="section__description"
          >
            <p
              class="text text--sm section__text"
            >
              This is a description of what the section is
            </p>
          </div>
        </div>
      </div>
    </header>
    <div
      class="section__body"
    >
      This is the section body, where you can put any content or include other components.
    </div>
  </div>
</section>
`;

exports[`<Section /> WithOverline story renders snapshot 1`] = `
<section
  class="section"
>
  <div
    class="section__inner"
  >
    <header
      class="section__header"
    >
      <div
        class="section__header-inner"
      >
        <div
          class="section__title-inner"
        >
          <div
            class="section__overline"
          >
            <p
              class="text text--overline"
            >
              Overline above title
            </p>
          </div>
          <h2
            class="heading heading--size-h2 section__title"
          >
            Section Title
          </h2>
        </div>
      </div>
    </header>
    <div
      class="section__body"
    >
      This is the section body, where you can put any content or include other components.
    </div>
  </div>
</section>
`;

exports[`<Section /> WithRight story renders snapshot 1`] = `
<section
  class="section"
>
  <div
    class="section__inner"
  >
    <header
      class="section__header"
    >
      <div
        class="section__header-inner"
      >
        <div
          class="section__title-inner"
        >
          <h2
            class="heading heading--size-h2 section__title"
          >
            Section Title
          </h2>
        </div>
      </div>
      <div
        class="section__right"
      >
        <button
          class="clickable-style clickable-style--md clickable-style--primary clickable-style--brand button button--primary"
          type="button"
        >
          Button
        </button>
      </div>
    </header>
    <div
      class="section__body"
    >
      This is the section body, where you can put any content or include other components.
    </div>
  </div>
</section>
`;

exports[`<Section /> WithTitleAfter story renders snapshot 1`] = `
<section
  class="section"
>
  <div
    class="section__inner"
  >
    <header
      class="section__header"
    >
      <div
        class="section__header-inner"
      >
        <div
          class="section__title-inner"
        >
          <h2
            class="heading heading--size-h2 section__title"
          >
            Section Title
            <span
              class="section__title-after"
            >
              <svg
                class="icon"
                fill="currentColor"
                height="1.375rem"
                role="img"
                style="--icon-size: 1.375rem;"
                viewBox="0 0 24 24"
                width="1.375rem"
                xmlns="http://www.w3.org/2000/svg"
              >
                <title>
                  help
                </title>
                <path
                  d="M11.95 18C12.3 18 12.5958 17.8792 12.8375 17.6375C13.0792 17.3958 13.2 17.1 13.2 16.75C13.2 16.4 13.0792 16.1042 12.8375 15.8625C12.5958 15.6208 12.3 15.5 11.95 15.5C11.6 15.5 11.3042 15.6208 11.0625 15.8625C10.8208 16.1042 10.7 16.4 10.7 16.75C10.7 17.1 10.8208 17.3958 11.0625 17.6375C11.3042 17.8792 11.6 18 11.95 18ZM12 22C10.6167 22 9.31667 21.7375 8.1 21.2125C6.88333 20.6875 5.825 19.975 4.925 19.075C4.025 18.175 3.3125 17.1167 2.7875 15.9C2.2625 14.6833 2 13.3833 2 12C2 10.6167 2.2625 9.31667 2.7875 8.1C3.3125 6.88333 4.025 5.825 4.925 4.925C5.825 4.025 6.88333 3.3125 8.1 2.7875C9.31667 2.2625 10.6167 2 12 2C13.3833 2 14.6833 2.2625 15.9 2.7875C17.1167 3.3125 18.175 4.025 19.075 4.925C19.975 5.825 20.6875 6.88333 21.2125 8.1C21.7375 9.31667 22 10.6167 22 12C22 13.3833 21.7375 14.6833 21.2125 15.9C20.6875 17.1167 19.975 18.175 19.075 19.075C18.175 19.975 17.1167 20.6875 15.9 21.2125C14.6833 21.7375 13.3833 22 12 22ZM12 20C14.2333 20 16.125 19.225 17.675 17.675C19.225 16.125 20 14.2333 20 12C20 9.76667 19.225 7.875 17.675 6.325C16.125 4.775 14.2333 4 12 4C9.76667 4 7.875 4.775 6.325 6.325C4.775 7.875 4 9.76667 4 12C4 14.2333 4.775 16.125 6.325 17.675C7.875 19.225 9.76667 20 12 20ZM12.1 7.7C12.5167 7.7 12.8792 7.83333 13.1875 8.1C13.4958 8.36667 13.65 8.7 13.65 9.1C13.65 9.46667 13.5375 9.79167 13.3125 10.075C13.0875 10.3583 12.8333 10.625 12.55 10.875C12.1667 11.2083 11.8292 11.575 11.5375 11.975C11.2458 12.375 11.1 12.825 11.1 13.325C11.1 13.5583 11.1875 13.7542 11.3625 13.9125C11.5375 14.0708 11.7417 14.15 11.975 14.15C12.225 14.15 12.4375 14.0667 12.6125 13.9C12.7875 13.7333 12.9 13.525 12.95 13.275C13.0167 12.925 13.1667 12.6125 13.4 12.3375C13.6333 12.0625 13.8833 11.8 14.15 11.55C14.5333 11.1833 14.8625 10.7833 15.1375 10.35C15.4125 9.91667 15.55 9.43333 15.55 8.9C15.55 8.05 15.2042 7.35417 14.5125 6.8125C13.8208 6.27083 13.0167 6 12.1 6C11.4667 6 10.8625 6.13333 10.2875 6.4C9.7125 6.66667 9.275 7.075 8.975 7.625C8.85833 7.825 8.82083 8.0375 8.8625 8.2625C8.90417 8.4875 9.01667 8.65833 9.2 8.775C9.43333 8.90833 9.675 8.95 9.925 8.9C10.175 8.85 10.3833 8.70833 10.55 8.475C10.7333 8.225 10.9625 8.03333 11.2375 7.9C11.5125 7.76667 11.8 7.7 12.1 7.7Z"
                />
              </svg>
            </span>
          </h2>
        </div>
      </div>
    </header>
    <div
      class="section__body"
    >
      This is the section body, where you can put any content or include other components.
    </div>
  </div>
</section>
`;

exports[`<Section /> WithTitleBefore story renders snapshot 1`] = `
<section
  class="section"
>
  <div
    class="section__inner"
  >
    <header
      class="section__header"
    >
      <div
        class="section__header-inner"
      >
        <div
          class="section__title-before"
        >
          <svg
            class="icon"
            fill="currentColor"
            height="1.375rem"
            role="img"
            style="--icon-size: 1.375rem;"
            viewBox="0 0 24 24"
            width="1.375rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              help
            </title>
            <path
              d="M11.95 18C12.3 18 12.5958 17.8792 12.8375 17.6375C13.0792 17.3958 13.2 17.1 13.2 16.75C13.2 16.4 13.0792 16.1042 12.8375 15.8625C12.5958 15.6208 12.3 15.5 11.95 15.5C11.6 15.5 11.3042 15.6208 11.0625 15.8625C10.8208 16.1042 10.7 16.4 10.7 16.75C10.7 17.1 10.8208 17.3958 11.0625 17.6375C11.3042 17.8792 11.6 18 11.95 18ZM12 22C10.6167 22 9.31667 21.7375 8.1 21.2125C6.88333 20.6875 5.825 19.975 4.925 19.075C4.025 18.175 3.3125 17.1167 2.7875 15.9C2.2625 14.6833 2 13.3833 2 12C2 10.6167 2.2625 9.31667 2.7875 8.1C3.3125 6.88333 4.025 5.825 4.925 4.925C5.825 4.025 6.88333 3.3125 8.1 2.7875C9.31667 2.2625 10.6167 2 12 2C13.3833 2 14.6833 2.2625 15.9 2.7875C17.1167 3.3125 18.175 4.025 19.075 4.925C19.975 5.825 20.6875 6.88333 21.2125 8.1C21.7375 9.31667 22 10.6167 22 12C22 13.3833 21.7375 14.6833 21.2125 15.9C20.6875 17.1167 19.975 18.175 19.075 19.075C18.175 19.975 17.1167 20.6875 15.9 21.2125C14.6833 21.7375 13.3833 22 12 22ZM12 20C14.2333 20 16.125 19.225 17.675 17.675C19.225 16.125 20 14.2333 20 12C20 9.76667 19.225 7.875 17.675 6.325C16.125 4.775 14.2333 4 12 4C9.76667 4 7.875 4.775 6.325 6.325C4.775 7.875 4 9.76667 4 12C4 14.2333 4.775 16.125 6.325 17.675C7.875 19.225 9.76667 20 12 20ZM12.1 7.7C12.5167 7.7 12.8792 7.83333 13.1875 8.1C13.4958 8.36667 13.65 8.7 13.65 9.1C13.65 9.46667 13.5375 9.79167 13.3125 10.075C13.0875 10.3583 12.8333 10.625 12.55 10.875C12.1667 11.2083 11.8292 11.575 11.5375 11.975C11.2458 12.375 11.1 12.825 11.1 13.325C11.1 13.5583 11.1875 13.7542 11.3625 13.9125C11.5375 14.0708 11.7417 14.15 11.975 14.15C12.225 14.15 12.4375 14.0667 12.6125 13.9C12.7875 13.7333 12.9 13.525 12.95 13.275C13.0167 12.925 13.1667 12.6125 13.4 12.3375C13.6333 12.0625 13.8833 11.8 14.15 11.55C14.5333 11.1833 14.8625 10.7833 15.1375 10.35C15.4125 9.91667 15.55 9.43333 15.55 8.9C15.55 8.05 15.2042 7.35417 14.5125 6.8125C13.8208 6.27083 13.0167 6 12.1 6C11.4667 6 10.8625 6.13333 10.2875 6.4C9.7125 6.66667 9.275 7.075 8.975 7.625C8.85833 7.825 8.82083 8.0375 8.8625 8.2625C8.90417 8.4875 9.01667 8.65833 9.2 8.775C9.43333 8.90833 9.675 8.95 9.925 8.9C10.175 8.85 10.3833 8.70833 10.55 8.475C10.7333 8.225 10.9625 8.03333 11.2375 7.9C11.5125 7.76667 11.8 7.7 12.1 7.7Z"
            />
          </svg>
        </div>
        <div
          class="section__title-inner"
        >
          <h4
            class="heading heading--size-h4 section__title"
          >
            Section Title
          </h4>
        </div>
      </div>
    </header>
    <div
      class="section__body"
    >
      This is the section body, where you can put any content or include other components.
    </div>
  </div>
</section>
`;
